### ISSUE TITLE
Statistics for buttons

### ISSUE THREAD

**Issue by irilyth** [buttonmen-dev/buttonmen#614]
It'd be cool if there were a way to see statistics for buttons, e.g. win/loss records vs other buttons. Even fancier stuff could be nice too, like a way to see the W/L/T numbers for the games played between two buttons, or even to actually see the games themselves (if they're public or yours), etc.

**Comment by cgolubi1** [comment]
Designating this as my BotM for February.  Next steps are to populate the DHS data on the dev and prod sites, and start working on API methods to access the data.

**Comment by blackshadowshade** [comment]
So, just noting that there is some progress on generating stats from the new site in Issue #2208.

**Comment by blackshadowshade** [comment]
My first proof-of-concept attempt to generate stats (based on my local branch 614_r_stats) produces a HTML page which can be found here: http://stats.dev.buttonweavers.com/ui/stats/button_stats.html

Comments are welcome.

Things I've thought of already:
- adding new columns, e.g., is the button TL, is the button a Fanatic, recipe
- allowing sorting on columns or other dynamic behaviour (which I would consider doing using a Shiny server)
- converting names to links

**Comment by irilyth** [comment]
Is the plan to generate stats into HTML pages, rather than into data in the databse?

**Comment by blackshadowshade** [comment]
I think that a mixture of the two approaches might be appropriate. Some stats pages that are commonly consulted are probably better suited to be static HTML pages, while customised stats pages (e.g., those pertaining to a particular user) may need to be dynamically generated from supporting summary data stored in the database.

We know from the old site just how much server load statistics pages can generate, and don't want to end up copping out and adding the standard disclaimer "please don't do this more than 10 times a day", especially if static HTML pages are enough for most purposes.

**Comment by dwvanstone** [comment]
Thank you! This is wonderful!

Are the buttons (if any) that have played 0 games in this document too?

It would be great to have the Button Set name as another column, especially if it becomes sortable. 

There are some buttons at the top I don't recognize from the production site: Skinny & Gordo.

**Comment by blackshadowshade** [comment]
No, buttons that haven't played any games (apart from mirror battles) aren't included at the moment.

Game 432 has Skinny, and game 416 has Gordo. This must have been before we disabled those buttons where we didn't support the full recipe.

**Comment by irilyth** [comment]
Here are some things I'd like to do with button vs button stats.

- Show an overall button ranking page, probably similar to the player ranking page. The particular thing I care about for my purposes (Mad Robin and UBFC) is overall win percentage, with a minimum number of total games played, ideally a user-configurable one.

- Something very fancy (which could also be used for player rankings, but my immediate interest is for button rankings) would be to consider the win percentage of each button's opponents in its ranking -- a "strength of schedule" adjustment. A button's strength-adjusted score might be its basic win percentage times the mean per-game win percentage of its opponents; I can get more into a couple of ways to calculate this if it's not entirely obvious. For extra bonus fun, allow this to go multi-level, so you could adjust each button's strength-adjusted score based on the strength-adjusted score of its opponents, to get a double-strength-adjusted score.

- For Mad Robin, and for general adoption purposes, it'd be cool if you could quickly identify buttons that a button has never played, or not played five times. For general adoption purposes, this could let you do things like create a game between a button and a random button that it hasn't played five times.

- What I really want for Mad Robin is a way to say "start with this button; then find seven other buttons who it has never played (or played fewer than five times), and who have never played (or played fewer than five times) each other, starting with the buttons closest to it in ranking, and expanding up and down evenly". I have a script that does this now, but it does it by reading the old site's stats from a gigantic JSON file on my local disk, so it's pretty fast. What that file actually contains is pairings: The file says that button X's record against Y is N - M, so I can just suck that into the data structure I want for my purposes, and then do my logic from there. That would be prohibitively expensive to get from the live site via the History API, since I'd need an API call for every pair of buttons... So a stats way to get one giant precomputed dataset for all the buttons at once would be super cool; or, even, a way to get all the pairings for one button at a time.

- For Mad Robin, UBFC, and adoption purposes, having the old-site stats integrated.

Just some ideas. Yay stats!

**Comment by blackshadowshade** [comment]
Regarding data formats, since what you're looking to do is import data so that you can do further processing, would CSV be enough?

**Comment by irilyth** [comment]
I think I care about the contents of the data much more than the format -- doesn't matter at all to me if it's CSV, JSON, whatever, as long as it's something easy for Python to parse. JSON is particularly easy, but CSV is probably doable, if that's easier to generate for some reason.

**Comment by blackshadowshade** [comment]
The reason I ask about CSV vs JSON is this: CSV produces a multiline file, while JSON produces a single string. In this case, because there are so many matchups, the string ends up being extremely long, so I wonder whether this is preferable to a multiline file.

**Comment by cgolubi1** [comment]
I've only sort of been following this discussion, but you can pretty-print JSON so it shows up on multiple lines.  Does that help with anything?

**Comment by irilyth** [comment]
I don't care one tiny bit whether the output has line breaks or not.

**Comment by irilyth** [comment]
To clarify, I don't mean that in a dismissive way. :^) The only thing I can think of is that more line breaks would increase the size or the dataset, so having fewer lines might make the data transfer faster. But I bet the difference would be totally undetectable.

**Comment by blackshadowshade** [comment]
@irilyth, I've put up a 32 Mb JSON file for you:

http://stats.dev.buttonweavers.com/ui/stats/win_percentage_stats.json

(b1 is the button of interest, b2 is the opponent's button, wp is the win percentage, ng is the number of games played)

Is this roughly what you're looking for?

**Comment by blackshadowshade** [comment]
#2207 has a number of colours that may be useful for the future.

**Comment by irilyth** [comment]
> @irilyth, I've put up a 32 MB JSON file for you:
> 
> http://stats.dev.buttonweavers.com/ui/stats/win_percentage_stats.json
> 
> (b1 is the button of interest, b2 is the opponent's button, wp is the win percentage, ng is the number of games played)
> 
> Is this roughly what you're looking for?

I never got around to looking at this, but might, before the next season of Mad Robin!

A quick question: Are you re-generating this file when you generate stats each month? Or was it a one-time thing? If the latter, can you do it again as another one-time thing?

**Comment by blackshadowshade** [comment]
I regenerate all the files each time I run the stats, including this one.

Chaos is currently working on getting the R script running once a day, so you will have daily up-to-date stats when that's working.

**Comment by PoshFrosh** [comment]
Something I've thought of with regards to stats is how stats are affected when we make changes to the site?
To know this we would have to compare stats from before the change to stats after the change. Therefore we'd need to be able to slice the stats by date.

For instance, how different will Pappy vs Bruno stats be now that we've implemented the extra die special rule? It would be cool to be able to compare (after enough games have been played with the new special in place).

This is super NOT important (how useful would it really be?). It's just something I thought of while reading this issue and figured I'd share. It might be better suited to creating a one-off bespoke report when notable changes are made, rather than a type of live data set as discussed above.

**Comment by blackshadowshade** [comment]
I agree that one-off bespoke reports are probably the best way to handle this case.

**Comment by cgolubi1** [comment]
Okay, so i've done some futzing around with getting R running on a docker container, so far doing everything manually to figure out what's needed before puppetizing it.

Starting with the good news: i'm able to load a copy of the staging DB on my container and run stats on it, and generate stats output files.  Are they correct?  Who knows, but they look plausible (people play a lot of CustomBM games on staging...) and validating beyond that is out of scope for this current experiment.

Now:
* The anecdote that made me realise maintaining R in this way is likely to be a problem
* A couple of specific issues i have with R and package management
* Proposed solutions / directions

So, the anecdote:
* background:
 * R is one of those languages that has its own internal ecosystem for installing packages (this is fairly common)
 * In the Unix world, this involves building things from source code (this feels like a throwback to me, and is going to make building the R-including container image pretty slow and painful (e.g. on the order of 30 minutes to build the dependencies the stats script needs), but i'm not currently aware of any alternatives, so i'm putting a pin in that)
 * Installation progresses via a pattern where you name the package you want and the repo you want to get it from, and you get the most recent version of that package that that repo has, and the most recent versions of all of that package's uninstalled dependencies, e.g. `install.packages('RMySQL', repos='https://cran.r-project.org')` --- worth noting, not only is the default not to specify a version, you actually *can't* specify a version, that function doesn't support it.
  * If you've discovered a specific older version of a package you want to install, you can install it, albeit i think without dependency support (`install.packages(packageurl, repos=NULL, type="source")`), but that's not helpful for any actual situation here, so i'm just mentioning that to indicate that i know it and so you don't need to tell me it.
* situation:
 * I installed R 3.2.3, which is the default version on our OS, and tried to install the dependencies, and got stuck on: `package 'data.table' is not available (for R version 3.2.3)`.  This is true: https://cran.r-project.org/web/packages/data.table/index.html says version 1.16.4 (currently the latest, though if you come back to this post in more than 10 minutes, that link will have bitrotted) depends on `R (≥ 3.3.0)`.
 * So the first thing i tried was going into the [package version archive](https://cran.r-project.org/src/contrib/Archive/data.table/) - and picking the latest pre-1.16 version, [data.table 1.15.4](https://cran.r-project.org/src/contrib/Archive/data.table/data.table_1.15.4.tar.gz) and seeing if that happened to support R 3.2.3.  As it turns out, it did.  So i felt good for like 13 seconds.
 * Then i discovered that the latest version of [dplyr](https://cran.r-project.org/web/packages/dplyr/index.html) which has a complex C++ compile and a zillion of its own dependencies, also doesn't work with 3.2.3.  At that point, i decided trying to move forward with 3.2.3 was going to be too much of a rathole, and cut bait on it.
 * I was then able to install R 4.4.2 by bringing in another (Ubuntu) repo (which i did in a somewhat untidy way for testing, and i'll have to clean that up before having something commitable, but anyway it works), and it seems to be possible to build all the dependencies under that version (at this moment in time), and with those dependencies installed and a couple of other script modifications, it seems to be possible to run the script and create stats, as noted.
* I think this anecdote is really worth keeping in mind, because James was confident the script should run under 3.2.3, and i'm sure that's right --- i'm sure the script runs fine, it's just that R's "always take the latest" attitude towards dependency installs, means that what you can install today, it's very likely you can no longer install tomorrow.  And that's no good for our infrastructure

The second thing i want to flag is really a side effect of my debug path on the first thing: at one point, i was trying to figure out how the actual code in `calcStats.R` uses `dplyr`, and i realised that i couldn't easily do so, because module imports in R bring all functions into the global namespace (R terminology is that they are "attached" to the global namespace), and i'm sure (kind of sure) there are introspection tools that can tell you where the current function you have is found in the search path, but i'd need to do more research about that.  And that lack of knowledge seems to me to make reviews hard, because it's not necessarily totally clear what effect a given diff has on which dependencies are needed or what they're used for.  (I also discovered in the process of doing research on solutions, that R silently and quasi-arbitrarily (latest import wins) resolves name conflicts, so if you use `filter()`, you can't necessarily know with certainty whether you are getting `dplyr::filter` or `stats::filter`, so that's not great.)

Anyway, so the two problems i want to propose solutions for are:
* How do we manage dependency versions portably?
* How do we maintain within the code what dependencies are actually being used and for what?

I found [this reddit thread](https://www.reddit.com/r/datascience/comments/ywg9ur/does_anyone_feel_like_r_is_actually_vastly_worse/) pretty helpful in getting a sense of what the options are and what kinds of approaches people are taking.

From that, it seems like "use renv" is the accepted answer for the first problem --- basically that's a mechanism to record what version of each dependency to install, in a lockfile that can be committed to git.  https://rstudio.github.io/renv/articles/renv.html is a good intro to what it does, and it looks like exactly what we want.  As that notes, it means we'll own a maintenance task of updating our dependencies periodically, but that's way better than not being able to communicate in code about what versions of things we're respectively using.  So i will look into how to use that.

For the second problem, it seems like it's *possible* to write R in a more maintainable way, but it's not the default.  The first post i read was [this one](https://conflicted.r-lib.org/) which i feel like explained to me the problem and some possible helpful syntax, so that's great, but i don't think `conflicted` itself is the answer, because it only solves for name conflicts, and i want to solve for reviewability as well.  I'm currently playing with [box](https://github.com/klmr/box) which seems to be obtainable from the base CRAN repos (i know that page says to use the author's own repo, but i didn't need to).  `box` provides a drop-in replacement for `library()`, `box::use()`, which lets you bring in modules or libraries without magically attaching everything in them to the global namespace, so you still have to actually document your code.  So far, i've converted the RMySQL/DBI import to use box, and it was easy and i like it.  I'm going to keep playing with that.

For now, just an FYI about what i'm looking into, and let me know if you have concerns about either of these approaches.

**Comment by cgolubi1** [comment]
James, i'll wait to hear your take on the overall approach, but i've finished my box-ified version of `calcStats.R`, and verified via my run vs the staging database that the results are the same (only timestamps differ):

```
# diff -r /var/www/ui/stats{.bk,}
diff -r /var/www/ui/stats.bk/button_matchup_stats.html /var/www/ui/stats/button_matchup_stats.html
2c2
< <!-- Mon Jan  6 18:58:53 2025 -->
---
> <!-- Mon Jan  6 19:45:18 2025 -->
diff -r /var/www/ui/stats.bk/button_stats.html /var/www/ui/stats/button_stats.html
2c2
< <!-- Mon Jan  6 18:58:07 2025 -->
---
> <!-- Mon Jan  6 19:44:37 2025 -->
```


Here's the WIP code diff FYI: https://github.com/buttonmen-dev/buttonmen/compare/master...cgolubi1:buttonmen:614_calculate_stats#diff-19ca159f9b27af647b9ba2a0f3bb0ca423f9b106e656b55aef7bc08b8cebe022R6-R8

**Comment by blackshadowshade** [comment]
I have no problems with your diff conceptually. I'll try running it at home and see whether I have any issues there.

**Comment by blackshadowshade** [comment]
I can confirm that the boxified script works perfectly fine for me. I'm happy for that to be checked in, if that's useful to you.

**Comment by cgolubi1** [comment]
Now that we've upgraded Ubuntu, i believe this is the current set of manual commands which will install and run stats on our site container:

```
mkdir /var/www/ui/stats
apt-get install make gcc g++ libmariadb-dev r-base libuv1-dev

R --vanilla

install.packages('RMySQL', repos='https://cran.r-project.org')
install.packages('xtable', repos='https://cran.r-project.org') 
install.packages('data.table', repos='https://cran.r-project.org')
install.packages('dplyr', repos='https://cran.r-project.org')
install.packages('condformat', repos='https://cran.r-project.org')
install.packages('box', repos='https://cran.r-project.org')

R --vanilla < /buttonmen/tools/stats/R/calcStats.R
```

Note that this relies on the database access parameters which are hardcoded into `calcStats.R`, which aren't correct for our remote databases.  But the point of this comment is to show the latest version of the dependency packages for the infrastructure.
